common/reactive/src/main/java/io/helidon/common/reactive/MultiFromInputStream.java

```diff
@@ -0,0 +1,186 @@
+
+        InputStreamSubscription(Flow.Subscriber<? super ByteBuffer> downstream,
+                                InputStream inputStream,
+                                final int bufferSize,
+                                ExecutorService executorService) {
+            this.downstream = downstream;
+            this.inputStream = inputStream;
+            this.bufferSize = bufferSize;
+            this.executorService = executorService;
+        }
+
+
+        private void trySubmit(long n) {
+            long emitted = 0L;
+            Flow.Subscriber<? super ByteBuffer> downstream = this.downstream;
+
+            for (;;) {
+                while (emitted != n) {
+                    int isCanceled = canceled;
+                    if (isCanceled != 0) {
+                        var toClose = inputStream;
+                        inputStream = null;
+                        if (isCanceled == BAD_REQUEST) {
+                            downstream.onError(new IllegalArgumentException(
+                                    "Rule §3.9 violated: non-positive request amount is forbidden"));
+                        } else if (isCanceled == NORMAL_CANCEL) {
+                            try {
+                                toClose.close();
+                            } catch (IOException e) {
+                                // Rule §1.8 says eventually
+                                downstream.onError(e);
+                            }
+                        }
+                        return;
+                    }
+
+                    ByteBuffer value;
+
+                    try {
+                        value = ByteBuffer.wrap(inputStream.readNBytes(bufferSize));
+                    } catch (Throwable ex) {
+                        inputStream = null;
+                        canceled = NORMAL_CANCEL;
+                        downstream.onError(ex);
+                        return;
+                    }
+
+                    if (value.limit() == 0) {
+                        inputStream = null;
+                        downstream.onComplete();
+                        return;
+                    }
+
+                    downstream.onNext(value);
+
+                    if (canceled != 0) {
+                        continue;
+                    }
+
+                    emitted++;
+                }
+
+                n = get();
+                if (n == emitted) {
+                    n = SubscriptionHelper.produced(this, emitted);
+                    if (n == 0L) {
+                        return;
+                    }
+                    emitted = 0L;
+                }
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0L) {
+                canceled = BAD_REQUEST;
+                n = 1; // for cleanup
+            }
+
+            if (SubscriptionHelper.addRequest(this, n) != 0L) {
+                return;
+            }
+
+            if (executorService == null) {
```

Comment:
This approach is 
1. error prone
2. hard to read

please either use two implementation classes, or use an executor service that executes in thread.

Reply:
fixed